### ISSUE TITLE
write-protect BUG?

### ISSUE THREAD

**Issue by tigerzhou100800** [keirf/FlashFloppy#135]
With these settings FlashFloppyV0.9.22a write-protect always Enable,
"Press eject button for 2 second" doesn't Disable write-protect.
Act like "write-protect = yes" .
I have to short GND and RST once to make "write-protect = no"  actually work.
And next time connect to power , it needs to short GND and RST again.
FlashFloppy mod Gotek SFRC9220 with SDcard reader , write able in windows .
settings:
======================
 interface = ibmpc
host = unspecified

write-protect = no

side-select-glitch-filter = 0
track-change = instant
index-suppression = yes
ejected-on-startup = no
image-on-startup = init
display-probe-ms = 3000
autoselect-file-secs = 2
autoselect-folder-secs = 255
nav-mode = default
nav-loop = yes
twobutton-action = zero
rotary = half
display-type = oled-128x32-rotate
oled-font = 6x13
display-off-secs = 60
display-on-activity = yes
display-scroll-rate = 200
display-scroll-pause = 2000
nav-scroll-rate = 80
nav-scroll-pause = 300
step-volume = 20
da-report-version = ""
extend-image = yes

**Comment by keirf** [comment]
Your USB stick is reporting it is write protected at power on. Try a different stick, or try pulling and reinserting the stick.

**Comment by tigerzhou100800** [comment]
Replug doesn't work , only GND-RST works .
Since GND-RST works , why can't it set to "write-protect = no" at power on.
USB stick with "write-protect = no" FF.CFG had already plug in when power on.

**Comment by keirf** [comment]
FF never writes to a USB stick that reports itself read-only, regardless of FF configuration.

After GND-RST, if you replug the stick does it become read-only again? Probably it does. What make is it?

**Comment by tigerzhou100800** [comment]
Why USB stick that reports itself read-only , Plug to every computer , all files deletable writable.
It's writable on Windows 98se with Asus P2BF, writable on Windows XP with Asrock conroe865GV,
writable on Windows 7 with Asrock B85 Pro4. And so this USB stick shall be read-only, 
Those systems and USB port hardware must be something wrong.
And yes, After GND-RST  replug it got "write-protect" lock  on  "yes"  again.

**Comment by keirf** [comment]
I'm not sure that ignoring the reported write protect status is a good idea in general. I imagine this setup reports this status for a short while at startup and for some reason Windows probes it in a way that it gets the more correct answer. A lot of these devices are tested to work on Windows, rather than work to spec...

I could add a config option to ignore USB write-protect status, or perhaps an option to delay USB startup for some milliseconds to give the drive a chance to get into the better state you see after GND-RST?

**Comment by tigerzhou100800** [comment]
Maybe a SDcard reader with respect SDcard write-protect switch, would first report "write-protect on" when power on , but after check SDcard within itself , it would report write-protect state again according to 
the SDcard's write-protect switch state. Then FlashFloppy maybe ignore the second report state.

**Comment by Squirrel61** [comment]
Something in this issue tells me that mr. Zhou is a Chinese and the stick he's using is one of these infamous Chinese sticks that you can also find on eBay or AliExpress, that always report a higher capacity than the real capacity of the memory chip(s) used.

I ran into such sticks a few times and even on Windows, also got strange effects including the mysterious write-protect.

Mr. Zhou, can you post some more information on the stick? Maybe a picture or a link to the seller?

**Comment by keirf** [comment]
It's an unbranded TF card in a USB reader...

Really, the best option here is to just try another stick, or reader. It also gives some more info on the original problem and whether the reader or card is misreporting.

I could add a FF.CFG option to ignore USB wrpro status.

**Comment by tigerzhou100800** [comment]
TFcard was from mini projector 's dead body , it's OEM with out brand print on it .
SDcard-skin doesn't matter ,  it extends TFcard pins to SDcard pins.
SDcard reader http://en.ssk.cn/Product/Rcard/SD/12.html

**Comment by tigerzhou100800** [comment]
quote: "or perhaps an option to delay USB startup for some milliseconds to give the drive a chance to get into the better state you see after GND-RST?"

Or power on USB reader first , check USB reader status later.

**Comment by keirf** [comment]
Power to USB is not gated at all. 5v from floppy connector flows straight to Vbus USB.

**Comment by Squirrel61** [comment]
I also had my share of unbranded (and branded, but faked) TF-cards from China and although most worked very well, some of them showed similar behaviour. Write-protected for Windows, but still perfectly usable in R4-cards (an R4 is a kind of Gotek for the Nintendo DS ;) ). Many times they became totally unreadable after some extended use.

My advice is to use a genuine TF-card from a good brand (although in China, you never know for sure if it's genuine but if there's one of those oval gold-colored warranty labels on the back, it's definitely fake). That's easier than to reprogram FlashFloppy to accept crappy cards for the few hours of life they have left.

**Comment by tigerzhou100800** [comment]
TFcard works fine , IT NEVER MISTAKELY LOCK WRITE-PROTECTED IN WINDOWS.  And can operate fine with android phone with TFcard port. I think it's real SanDisk TF card.  
I think QUOTE: "an option to delay USB startup for some milliseconds" would be fine.

**Comment by keirf** [comment]
Here is a firmware which allows you to ignore the WP status of the attached USB drive. See how that works out for you, including that writes do actually work ;)

You need the following option in FF.CFG: `usb-ignore-wp = yes`

[ff_135_1.zip](https://github.com/keirf/FlashFloppy/files/2140083/ff_135_1.zip)

**Comment by tigerzhou100800** [comment]
Yes, it works. But, I just keep it old way, since I don't write to emulator floppy disk much , it's better be write protected . I add a button to GND-RST , it temporarily remove write-protection. Just forget about this thread .

**Comment by xsrf** [comment]
@keirf did you include this feature? I can't find `usb-ignore-wp` on https://github.com/keirf/flashfloppy/wiki/FF.CFG-Configuration-File
I do have the issue that all my img-Files are write protected even when I set write protect off in eject menu and when I go back to eject menu it's on again. I tried `usb-ignore-wp` just out of curiosity but it didn't help. I'm on latest v3.42.

**Comment by keirf** [comment]
I did not, but I just made a patch to always ignore USB write-protect status and pushed that to a test branch. You can download the build artifact here (if logged in to GitHub): https://github.com/keirf/flashfloppy/actions/runs/11870135955

Just try flashing the non-debug update file and see how it goes.

**Comment by xsrf** [comment]
@keirf thx, turns out I just don't know how to use flashfloppy. I have to exit Eject Menu on `**Eject Menu**` screen for WP to persist. Exiting via `Exit to selector` re-enables write protection. Opening Eject Menu again also re-enables write protection.
Not really obvious ...

**Comment by keirf** [comment]
Write protect toggle doesn't persist after the current image is unmounted. Because it doesn't get written back to usb. Hence exit to selector will lose your setting. Re-entering eject menu from currently mounted image should not forget the WP status. I will have to check that.

**Comment by xsrf** [comment]
> Write protect toggle doesn't persist after the current image is unmounted.
I got that, read that in few other issues before.
But to me it was not obvious that `Exit to selector` will unmout the image, nor that `**Eject Menu**` even is a valid item to select - It's a menu title to me...

**Comment by keirf** [comment]
There is also option `Exit & re-insert` which is in contrast with `Exit to selector`. You can use this option if you prefer not to use `**Eject Menu**` option, which is indeed just a convenience. You can find `Exit to selector` one position left from `**Eject Menu**`.

**Comment by keirf** [comment]
What you *really* want is probably what's described in #516 so go vote on that if so :)

**Comment by xsrf** [comment]
Well okay... Guess as a new user to the software this Menu is counter-intuitive to me...
`Exit & re-insert` implied to me that the disk is ejected first (why else "re-insert") so I didn't even try it, knowing that the option is not persisted in the image.
`Exit to selector` was the Option I thought would keep the image as it doesn't mention any actions. The "to selector" kinda irritated me though, because, where else would it exit to?

So basically:
Exit & re-insert = Exit but keep disk (mentions an action but does nothing)
Exit to selector = Exit & re-insert disk (mentions no action but performs an action)

Also, I noticed, just opening the eject menu already ejects the disk... Guess that explains the name, but I just expected a menu.

**Comment by keirf** [comment]
I have added the description of these operating modes to the wiki. It's kind of fundamental, and it wasn't described there at all. Now it's in "Initial Setup".

**Comment by xsrf** [comment]
That's good, but not really helpful. I don't get the difference between Navigation/Selector and Emulation. My Gotek with only its two buttons and default settings, from my perspective, can't be in three modes. I'm either in eject menu or in emulation. The Disk I select is emulated. There is no option to browse images but not mount them as there is no button or menu item to mount them explicitly. They are implicitly mounted by selection.
The `Exit to selector` option re-mounts the selected disk, as does navigating image files.
I don't know if it's different if you use the rotary encoder but from YT videos I watched it's not, as pressing the reotary encoder also just maps to the eject menu and not "mount selected image".

**Comment by keirf** [comment]
Currently highlighted image in navigation mode gets auto mounted (and thus switch to emulation mode) after two seconds. That's changeable via FF.CFG. 

The change between navigation/selector and emulation modes is subtle in terms of the display contents. You will see the track number appear when in emulation mode, instead of image type and directory depth.

Anyway, there isn't nornally much need to be able to distinguish between emulation and navigation modes as a user, given the usually automatic/implicit switching between them. But clearly it does interact mighty confusingly with non-persistent write-protect setting. Hence #516
